Further Performance improvements

VincentKueszter opened the pull request:
Hello,

I sat down and tried to get a bit more performance out of the script.
Baseline, vertex colors only: 23,8 fps
I made the following changes, one per commit:
- min/max clamping of weights per lambda: 24,08 fps (It's really hard to read imo, if this can be formatted in a better way, I'd be glad to improve it!)
- vertex_color as 2D data structure -> easier to read and 25,21 fps (with only this change)
- added global object for original edge length -> 25,8 fps (with only this change)
- added index mapping: from the loop-based index to a vertex-based index, this also simplifies the setting of the vertex colors -> 28,37fps (with only this change)

I now get 31,32fps with all those changes together.

I'll mark this as draft for now, since I want to look into it a bit more, but your input would be appreciated!

VincentKueszter committed:
clamping values by using lambda

VincentKueszter committed:
replacing vertex_colors with a 2D data structure

VincentKueszter committed:
added global object for tracking data to avoid recalculating it each frame

VincentKueszter committed:
added index map to improve and simplify setting the vertex colors

ScottishCyclops commented:
Thank you !
I'm about to read your code. That sounds promising.

ScottishCyclops commented on tensionmap.py:
```diff
@@ -155,6 +155,9 @@ def tm_update(obj, context):
     if obj.data.tm_enable_vertex_colors:
         vertex_colors = [0.0] * (number_of_tm_channels * num_vertices)
 
+    #lambda for clamping between min and max
```
Don't forget the space after the `#`
I don't know what IDE you use, but if you want to have the same auto formatting as me, I use Visual Studio Code with the `Python` extension, and the `autopep8` formater

```json
"python.formatting.provider": "autopep8"
```

If you use VS Code, I would also recommand `MagicPython` for development.

VincentKueszter replied:
Fixed.
I actually edit it in Blender, since it's quicker to just save/reload/check performance.
I always wanted to just switch to PyCharm, but I have pushed it off until now...

ScottishCyclops replied:
I personally have Blender and the User Prefs always opened, then when I update the code in VSCode, I delete the addon, then click install again, and I am presented with always the same folder that only contains the addon, then I enable it and it's done.
So it's relatively fast.

ScottishCyclops reviewed (commented):
> It's really hard to read imo, if this can be formatted in a better way, I'd be glad to improve it!

I've always found the python way of writing inline ifs (ternary) very readable, as in you can read the line of code in English and it makes sense.

No problem with readability in my opinion.

I'm glad to hear this increases performance, because it clearly increases the overall readability too :+1:

VincentKueszter committed:
Fixed comment formatting

ScottishCyclops commented on tensionmap.py:
```diff
@@ -153,7 +153,7 @@ def tm_update(obj, context):
     # create vertex color list for faster access only if vertex color is activated
     if obj.data.tm_enable_vertex_colors:
-        vertex_colors = [0.0] * (number_of_tm_channels * num_vertices)
+        vertex_colors = [[0.0]*number_of_tm_channels]* num_vertices
```
I would put spaces around the `*` operators. For consistency.

VincentKueszter replied:
Done!

ScottishCyclops commented on tensionmap.py:
```diff
@@ -179,10 +179,8 @@ def tm_update(obj, context):
-            # green
-            vertex_colors[i * number_of_tm_channels + 1] = squeeze_value
+            # red, green, blue, alpha
+            vertex_colors[i] = (stretch_value,squeeze_value, 0.0, 1.0)
```
Do use the formatter I was talking about in my other comment. Because here also, it is missing a space between the first two elements of the tuple.

I don't like nit picking about code styling yet I like it to be consistent, that's why I prefer having one formatter dictate the format of the whole code and never worry about it again.

I'll add in the README the formatter I use so that contributors know what style is expected.

VincentKueszter replied:
Yeah, I'll add a formatter.
Style has to be consistent and having to check and correct it manually each time is essentially wasting both of our time. :)

ScottishCyclops reviewed (commented):
Very nice, and a lot more readable again ! Good job.

ScottishCyclops commented on tensionmap.py:
```diff
@@ -221,6 +244,12 @@ def tm_update_selected(self, context):
+    if context.object.data.tm_active:
+        object_data[context.object.name]=Object_Data(context.object)
+    else:
+        del object_data[context.object.name]
```
Since you are using `context.object.name` no matter what, putting it in a variable can make the code more readable, and depending on how Python and the Blender API work, maybe also a bit faster.

`context.object` could also be put in a variable. I'm okay with either options here.

VincentKueszter replied:
Done!

ScottishCyclops commented on tensionmap.py:
```diff
@@ -43,6 +44,23 @@
 
 tm_update_modes = ["OBJECT", "WEIGHT_PAINT", "VERTEX_PAINT"]
 
+class Object_Data(bpy.types.Object):
```
I think the convention for Classes would be `ObjectData`.

I don't really like the name though. It doesn't really say what it does.

VincentKueszter replied:
I'm open for suggestions!

ScottishCyclops replied:
Sorry, I was supposed to write a suggestion, then I got carried away in my review xD

`TmCachedData` ?

Also does it have to inherit the `bpy.types.Object` ? Real question here.

VincentKueszter replied:
I had 2 reactions:
- Wait, this is how Python does inheritance?
- Why on earth did I write this? 😄 Probably mixed it up with C++ Constructors. The __init__ function suffices.
I won't fix it for now, I'll redo this block when implementing the caching you mentioned.

ScottishCyclops commented on tensionmap.py:
```diff
@@ -43,6 +44,23 @@
+        return original_edge_lengths
+
+    @property
+    def original_edge_length(self): 
```
I'm afraid I don't really follow the logic here. It doesn't help that there is  member called `original_edge_length**s**` with an s, and a method tagged as property named `original_edge_length` without s.

VincentKueszter replied:
Oh, the property is a typo, it should be `original_edge_lengths` in all cases.
The __calculate_original_edge_lengths function is just a helper function for initialization, but I guess the return value is superfluous, since we have access to self and could fill the variables inside the function, right?
As I mentioned, I am new to Python, so if something is not Pythonian, feel free to make suggestions!
PS: GitHub seems to have a "Insert a suggestion" feature, maybe we can try that here?

VincentKueszter replied:
I'll delay fixing this for now due to your other comments.

ScottishCyclops replied:
Maybe that object could be replaced with a named tuple. See https://stackoverflow.com/a/36033
And the two init functions for each values would become their own pure function, called when creating a new named tuple. (I like pure functions).

VincentKueszter commented:
> > It's really hard to read imo, if this can be formatted in a better way, I'd be glad to improve it!
> 
> I've always found the python way of writing inline ifs (ternary) very readable, as in you can read the line of code in English and it makes sense.
> 
> No problem with readability in my opinion.
> 
> I'm glad to hear this increases performance, because it clearly increases the overall readability too 👍

If I could, I would format it like this. It's a bit easier to see what happens in each case at a glance, in my opinion.
```
lower if value < lower else 
upper if value > upper else 
value
```

ScottishCyclops reviewed (commented):
I don't really like using a class for a single, may I say constant, variable.
I think the function (as it was in your previous commit) is better.

I'm the kind of dev that likes to use OOP only when it really makes sense, and keep to functions and data structures for the rest.

I don't think it resolves the problem of "If the object's geometry changes, you have to refresh manually", right ?

Here is what I propose:
This is an extra boost of real-time performance, for the ones that really need it. So why not use the same principle as the "cache" functionality found in many places in Blender ?
You would manually choose to "cache" a mesh for TensionMap, enabling this performance gain.
But since you manually chose to enable cache, then it's okay to have you update the cache when you change the mesh.

That way, most users have the current behaviour of auto updating every frame, but don't have the hassle of updating the mesh in TensionMap manually everytime.
And the hardcode real-time users can choose the functionality, knowing they will need to manually update.

If you don't want to work on the UI part of this, let me know, I'll gladly do it.
I'll take your current work here, turn it into a function, and link it to the UI.

VincentKueszter committed:
minor formatting changes

ScottishCyclops reviewed (commented):
Ok I like this.

But it would be under the same optional "cache" behaviour, because again, when you change your mesh, you would need to manually update the TensionMap. See my other comments.

VincentKueszter committed:
more temporary variables

ScottishCyclops commented:
> > > It's really hard to read imo, if this can be formatted in a better way, I'd be glad to improve it!
> > 
> > 
> > I've always found the python way of writing inline ifs (ternary) very readable, as in you can read the line of code in English and it makes sense.
> > No problem with readability in my opinion.
> > I'm glad to hear this increases performance, because it clearly increases the overall readability too +1
> 
> If I could, I would format it like this. It's a bit easier to see what happens in each case at a glance, in my opinion.
> 
> ```
> lower if value < lower else 
> upper if value > upper else 
> value
> ```

One line is fine by me. It starts to take more room than it really should otherwise.

VincentKueszter commented:
> Here is what I propose:
> This is an extra boost of real-time performance, for the ones that really need it. So why not use the same principle as the "cache" functionality found in many places in Blender ?
> You would manually choose to "cache" a mesh for TensionMap, enabling this performance gain.
> But since you manually chose to enable cache, then it's okay to have you update the cache when you change the mesh.
> That way, most users have the current behaviour of auto updating every frame, but don't have the hassle of updating the mesh in TensionMap manually everytime.
> And the hardcode real-time users can choose the functionality, knowing they will need to manually update.

Caching is a really fitting name for it! Sounds good.

> If you don't want to work on the UI part of this, let me know, I'll gladly do it.
> I'll take your current work here, turn it into a function, and link it to the UI.

I'll try it myself next week, the UI part didn't seem too complicated.

ScottishCyclops commented:
> I'll try it myself next week, the UI part didn't seem too complicated.

Alright.
I'm picturing a button that says "Cache Geometry Data" with a hover text of something like "Improve realtime performance by pre-calculating some geometry data. An update to the mesh will require a manual cache update".

When the cache is enabled, the button is replaced by two buttons (in the same line maybe) that say "Disable cache" and "Update Cache".

The update of the Cache would need to stop being linked with `tm_update`, because it would have it's own operator.

Have fun coding :)

VincentKueszter committed:
Replaced precalculation with Caching option - still calculates automatically each frame if caching is disabled.

VincentKueszter commented:
Okay, I re-implemented it as a cache.
I used dataclass instead of named tuple because it was a tiny bit faster.
I also updated the GUI and GUI code a bit, the row/col variable names were swapped and I removed some duplicated strings.
There should be no further performance improvements, even though I added a few more variables to avoid dot-chains.
Hope I didn't miss anything!

VincentKueszter committed:
Further performance improvements
using BMesh for faster edge length calculation
made clamp use def, not assignment as per E731
fix: channel size set to 4, not 2
removed a variable when setting colors_tension_data for a minor performance boost

VincentKueszter commented:
I added some more performance improvements by using bMesh:
7,9% more fps with vertex colors only and 10,2% with vertex groups only.

ScottishCyclops commented:
Hey, Haven't had much time recently. I'll be sure to test & merge your changes as soon as I can. Thanks for the contributions !

VincentKueszter commented:
Hey, did you have time to look over it yet?

winktim commented:
Hey, no sorry. Back to school here x)
This has gotten down my priority list again.

ScottishCyclops commented:
Hey,
No sorry. Back to school and haven't had time.
This has gotten down in my priority list.

VincentKueszter commented:
Alright! I'll wait for a notification then.
